keps/sig-node/4191-split-image-filesystem/README.md

```diff
@@ -0,0 +1,954 @@
+However, there are a few complications with this goal.
+
+1. E2E tests around eviction with a single disk are currently [CRI-O-eviction](https://testgrid.k8s.io/sig-node-cri-o#ci-CRI-O-cgroupv1-node-e2e-eviction) and [containerd-eviction](https://testgrid.k8s.io/sig-node-containerd#node-kubelet-containerd-eviction) failing.  
+
+2. There is zero test coverage around a separate image filesystem.  There is an issue to improve this at the unit test level.
+
+1 can be addressed by investigating the eviction tests and figure out the root cause of these failures.
+
+As part of this KEP, we should add testing around separate disks in upstream Kubernetes.  Since this is already a supported use case in Kubelet, there should be testing around this.
+
+Kubelet/CRI-O should be set up with configuration for a separate disk.  Eviction and Summary E2E tests should be added in the case of a separate disk.
+
+And tests for split image filesystem should be added.
+
+E2E Test Use Cases addition:
+
+- E2E tests for summary api with separate disk.
+  - Separate Disk - ImageFs reports separate disk from root when disk is mounted.
+  - Split Disk - Writeable layer on Node, read-only layer on ImageFs
+- E2E tests for eviction api with separate disk
+  - Replicate existing disk pressure eviction e2e tests with disk.
+
+### Graduation Criteria
+
+#### Alpha [Release the CRI API and Kubelet Changes]
+
+CRI API changes are composed in containerd and CRI-O so the CRI API must be released first.
+
+- Using `ImageFsInfo` is guarded with a feature gate.
+- Implementation for split image filesystem in Kubernetes.
+  - Eviction manager modifications in case of split filesystem.
+  - Summary and Stats Provider implementations
+- CRI API merged.
+- Unit tests
+- E2E tests to cover separate image filesystem
+  - It is not possible to have e2e tests for split filesystem at this stage.
+
+#### Alpha Part 2 [CRI-O, E2E Tests and CRITools]
+
+Shortly after this release and new CRI package, projects that consume the CRI API
+can be updated to use the new API features.
+
+- At least one CRI implementation supports split filesystem
+- E2E tests supporting the CRI implementation with split image filesystem
+- CRI-Tool changes for image fs
+
+#### Alpha To Beta Promotion
+
+- Gather feedback on other potential use cases.
+- Always set `KubeletSeparateDiskGC` to true so `ImageFsInfo` is used instead of `ImageStats` in all cases.
+- Always set `KubeletSeparateDiskGC` to true so that eviction manager will detect split file system and handle it correctly.
+
+#### Stable
+
+- More than one CRI implementation supports split filesystem
+
+### Upgrade / Downgrade Strategy
+
+There are two cases that this feature could impact users.
+
+Case 1: Turning the feature on with no split filesystem.
+In this case, the main difference will be that clusters that use the `CAdvisor Stats Provider`, we will switch to using `ImageFsInfo` to report
```

Comment:
feels that a mismatch between kubelet and container runtime (e.g. turning on this feature on kubelet but container runtime does not implement the CRI interface)might cause trouble for users. Is it possible for kubelet to detect whether containerd is ready for this feature, and error if not? 

For example, can kubelet check if `ImageFsInfo` returned from runtime is nil, and if so, then it probably means that the container runtime does not have necessary implementation for this feature, and in that case print and error and exit?

Reply:
So `ImageFsInfo` is implemented in both CRI-O and Containerd.  There is no danger of the endpoint returning nil entirely.  

The main issue will be if `container_filesystem` is returned.  In case of the containerd runtime not having this field populated, it would mean that (for alpha/alpha + 1), that imagefs=containerfs.

We can't really do what you suggest because it will fail a lot of the regression tests.  Most cases will not have this functionality and we will test against a containerd/CRI-O that is by default not set for splitting.  

I think the suggestion I have is probably the best way forward for version skew.  As we can't really assume that any container runtime implementation will have this feature and even if they do, we should correctly guard against configurations that do not allow this.